src/meta-srv/src/handler/failure_handler.rs

```diff
@@ -0,0 +1,151 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+mod runner;
+
+use api::v1::meta::HeartbeatRequest;
+use async_trait::async_trait;
+
+use crate::error::Result;
+use crate::handler::failure_handler::runner::{FailureDetectControl, FailureDetectRunner};
+use crate::handler::{HeartbeatAccumulator, HeartbeatHandler};
+use crate::metasrv::{Context, ElectionRef};
+
+#[derive(Eq, Hash, PartialEq, Clone)]
+pub(crate) struct RegionIdent {
+    catalog: String,
+    schema: String,
+    table: String,
+    region_id: u64,
+}
+
+// TODO(LFC): TBC
+pub(crate) struct DatanodeHeartbeat {
+    #[allow(dead_code)]
+    cluster_id: u64,
+    #[allow(dead_code)]
+    node_id: u64,
+    region_idents: Vec<RegionIdent>,
+    heartbeat_time: i64,
+}
+
+pub struct RegionFailureHandler {
+    failure_detect_runner: FailureDetectRunner,
+}
+
+impl RegionFailureHandler {
+    pub fn new(election: Option<ElectionRef>) -> Self {
+        Self {
+            failure_detect_runner: FailureDetectRunner::new(election),
+        }
+    }
+
+    pub async fn start(&mut self) {
+        self.failure_detect_runner.start().await;
+    }
+}
+
+#[async_trait]
+impl HeartbeatHandler for RegionFailureHandler {
+    async fn handle(
+        &self,
+        _: &HeartbeatRequest,
+        ctx: &mut Context,
+        acc: &mut HeartbeatAccumulator,
+    ) -> Result<()> {
+        if ctx.is_infancy {
+            self.failure_detect_runner
+                .send_control(FailureDetectControl::Purge)
+                .await;
+        }
+
+        if ctx.is_skip_all() {
+            return Ok(());
+        }
+
+        let Some(stat) = acc.stat.as_ref() else { return Ok(()) };
+
+        let heartbeat = DatanodeHeartbeat {
+            cluster_id: stat.cluster_id,
+            node_id: stat.id,
+            region_idents: stat
+                .region_stats
+                .iter()
+                .map(|x| RegionIdent {
+                    catalog: x.catalog.clone(),
+                    schema: x.schema.clone(),
+                    table: x.table.clone(),
+                    region_id: x.id,
+                })
+                .collect(),
+            heartbeat_time: stat.timestamp_millis,
+        };
```

Comment:
Is there a `From<State> for DatanodeHeartbeat` needed?

Reply:
Cannot `From<Stat>` because we don't own the `stat` here. Have to `From<&Stat>` which seems a little odd. I don't like creating new object from `From`, because the `into()`(as well as `try_into()`) has a hard time working with IDE: can't simply click into the method to see where the implementation really is. Besides, there's little rational relationship between `Stat` and `DatanodeHeartbeat`, so I prefer not use `From`.